adapters/tlsio_openssl.c

```diff
@@ -1116,57 +1274,71 @@ static const char *get_dp_url(DIST_POINT *dp)
     return NULL;
 }
 
-static X509_CRL *load_crl_crldp(X509 *cert, const char* suffix, STACK_OF(DIST_POINT) *crldp)
+static int is_valid_crl(X509 *cert, X509_CRL *crl)
 {
-    int i;
-    X509_CRL *crl = NULL;
-    char buf[256];
     time_t now = time(NULL);
+    X509_NAME *issuer_cert = X509_get_issuer_name(cert);
+
+    // names don't match up. probably a hash collision
+    // so lets test if there is another crl on disk.
+    X509_NAME *issuer_crl = X509_CRL_get_issuer(crl);
+    if (0 != X509_NAME_cmp(issuer_crl, issuer_cert))
+    {
+        return 0;
+    }
+
+    // Important: At this point, we will DELETE
+    //      a file holding a Crl from disk in case
+    //      the invalid-after date is less than the
+    //      current time.
+    //      This will trigger the re-loading of the
+    //      Crl from the download store, if available.
+    time_t crlend = crl_invalid_after(crl);
+    if (crlend <= now)
+    {
+        LogInfo("crl %ld, %ld outdated\n", crlend, now);
+        return 0;
+    }
+
+    return 1;
+}
+
+static int load_cert_crl_file(X509 *cert, const char* suffix, X509_CRL **pCrl)
+{
+    char buf[256];
+
+    int ret = 0;
+    *pCrl = NULL;
 
     char* prefix = NULL;
-    if( NULL == (prefix = getenv("TMP")) &&
+    if (NULL == (prefix = getenv("TMP")) &&
         NULL == (prefix = getenv("TEMP")) &&
         NULL == (prefix = getenv("TMPDIR")))
     {
-        prefix = ".";
+        return 0;
     }
 
     // we need the issuer hash to find the file on disk
     X509_NAME *issuer_cert = X509_get_issuer_name(cert);
     unsigned long hash = X509_NAME_hash(issuer_cert);
 
     // try to read from file
-    for (i = 0; i < 10; i++)
+    for (int i = 0; i < 10; i++)
     {
         sprintf(buf, "%s/%08lx.%s.%d", prefix, hash, suffix, i);
 
         // try to read from disk, exit loop, if
         // none found
-        crl = load_crl(buf, FORMAT_PEM);
+        X509_CRL* crl = load_crl(buf, FORMAT_PEM);
         if (!crl)
         {
             continue;
         }
 
-        // names don't match up. probably a hash collision
-        // so lets test if there is another crl on disk.
-        X509_NAME *issuer_crl = X509_CRL_get_issuer(crl);
-        if (0 != X509_NAME_cmp(issuer_crl, issuer_cert))
+        if (!is_valid_crl(cert, crl))
```

Comment:
aren't you leaking here crl? I mean when you load it you do reference++, I do not see where you do --/refdecrease.